src/coreclr/src/vm/dllimportcallback.cpp

```diff
@@ -961,12 +961,18 @@ void UMEntryThunk::Terminate()
     CONTRACTL
     {
         NOTHROW;
+        MODE_ANY;
     }
     CONTRACTL_END;
 
     m_code.Poison();
 
     s_thunkFreeList.AddToList(this);
+
+    if (GetObjectHandle())
```

Comment:
This needs to be done before `s_thunkFreeList.AddToList`. Once you add it to the freelist list, some other thread can get it out and work with it.